# ====================================================================================================================================
# @file       ethercat_driver.yaml
# @author     Krzysztof Pierczyk (krzysztof.pierczyk@gmail.com)
# @maintainer Krzysztof Pierczyk (krzysztof.pierczyk@gmail.com)
# @date       Tuesday, 22nd March 2022 3:12:52 am
# @modified   Tuesday, 31st May 2022 1:59:34 pm
# @project    engineering-thesis
# @brief      Default configuration for the velmwheel_ethercat_driver node
#    
#    
# @copyright Krzysztof Pierczyk Â© 2022
# ====================================================================================================================================

/**:

    # Parameters
    ros__parameters:

        # ========================================== Common parameters ======================================== #

        # -------------------------------------------------------------------------------------------------------
        # @details Description of the ENI configuration source. This shall be given as:
        #
        #    1) 3-element touple holding
        #  
        #           - source package
        #           - source resource
        #           - source file
        # 
        #        In such a case node assumes that the @p package registers ament resource named @p resource whose 
        #        content is a path to the directory containing ENI files (relative to the @package prefix). The 
        #        ENI file named @p file is expected to reside in this directory.
        #
        #    2) 1-element touple describing
        #
        #           - absolute path to the ENI file
        #
        #       In such a case node loads ENI file from the given path
        #
        # @note Given ENI file is not cached in the node's memory. It is reread every time the node need to 
        #    access ENI data (e.g. when loading slave driver). Location of the file must be valid for the entire
        #    node's lifetime
        # -------------------------------------------------------------------------------------------------------
        eni_source: [ 'velmwheel_twincat', 'eni', 'default_config.xml' ]

        # List of driver plugins that should be loaded on construction
        initial_drivers: []

        # ========================================== Process config =========================================== #
      
        # Configuration of the processing thread
        process:

            # Memory configruation
            memory:

              # -------------------------------------------------------------------------------------------------------
              # @brief Whether memory pages of the node process should be locked at node's initialization 
              #    (either 'none', 'current' or 'all'). When 'all' given current as well as future memory pages
              #    of the node will be locked.
              # -------------------------------------------------------------------------------------------------------
              lock: 'none'
              
              # If 'true' node will disable trimming memory with sbrk() at startup
              disable_trimming: false
              # If 'true' node will disable mmap() usage ats tartup
              disable_mmap: false

            # Configuration of the node thread
            node: 

                # Affinity of the node thread
                affinity: 0xF
                # Scheduling policy of the node thread (SCHED_OTHER, SCHED_FIFO, SCHED_RR)
                sched_policy: 'SCHED_OTHER'
                # Scheduling priority of the node thread
                sched_priority: 0
        
        # ====================================== CIFX Toolkit configuration =================================== #

        # CIFX configuration
        cifx:

            # Verbosity of the CIFX toolkit ('Debug', 'Info', 'Warning', 'Error')
            log_level: 'Info'
            # Polling interval of the IRQ-polling CIFX Toolkit thread in [s] (polling disabled if not given)
            # [uncomment to enable] cos_polling_interval: 0.001

        # Configruation of the CoS-polling thread
        cos_polling_thread:

            # Affinity of the IRQ-polling CIFX Toolkit thread
            affinity: 0xF
            # Scheduling policy of the IRQ-polling CIFX Toolkit thread (SCHED_OTHER, SCHED_FIFO, SCHED_RR)
            sched_policy: 'SCHED_OTHER'
            # Scheduling priority of the IRQ-polling CIFX Toolkit thread
            sched_priority: 0

        # Configruation of the IRQ thread
        irq_thread:

            # Affinity of the IRQ-handling CIFX Toolkit thread
            affinity: 0xF
            # Scheduling policy of the IRQ-handling CIFX Toolkit thread (SCHED_OTHER, SCHED_FIFO, SCHED_RR)
            sched_policy: 'SCHED_OTHER'
            # Scheduling priority of the IRQ-handling CIFX Toolkit thread
            sched_priority: 0
